let jsonData = {"mapping": {"groupName": "mapping", "modules": [{"name": "bwa-sw", "installed": true, "settings": {"required": [{"name": "genome_fasta", "label": "Reference Genome", "description": "Path to genome fasta file", "type": "file"}], "optional": [{"name": "additional_bwa_options", "label": "Additional BWA Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}, {"name": "bwa-mem", "installed": true, "settings": {"required": [{"name": "genome_fasta", "label": "Reference Genome", "description": "Path to genome fasta file", "type": "file"}], "optional": [{"name": "additional_bwa_options", "label": "Additional BWA Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}, {"name": "bowtie2", "installed": true, "settings": {"required": [{"name": "genome_fasta", "label": "Reference Genome", "description": "Path to reference genome fasta file", "type": "file"}, {"name": "alignment_type", "label": "Alignment Type", "description": "Choose local or end-to-end alignment", "type": "enum", "choices": {"local": "--very-sensitive-local", "end-to-end": "--very-sensitive"}}], "optional": [{"name": "additional_bowtie2_options", "label": "Additional Bowtie2 Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}, {"name": "bwa-backtrack", "installed": true, "settings": {"required": [{"name": "genome_fasta", "label": "Reference Genome", "description": "Path to genome fasta file", "type": "file"}], "optional": [{"name": "additional_bwa_options", "label": "Additional BWA Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}]}, "premapping": {"groupName": "premapping", "modules": [{"name": "fastqc", "installed": true}, {"name": "multiqc", "installed": true}]}, "preprocessing": {"groupName": "preprocessing", "modules": [{"name": "trimgalore", "installed": true, "settings": {"optional": [{"name": "phred_score_type", "label": "Phred Score Type", "description": "Choose phred33 (Sanger/Illumina 1.9+ encoding) or phred64 (Illumina 1.5 encoding)", "type": "enum", "default": "--phred33", "choices": {"phred33": "--phred33", "phred64": "--phred64"}}, {"name": "quality_threshold", "label": "Base-Quality Threshold", "description": "Trim low-quality ends from reads in addition to adapter removal. (Default: 20)", "type": "number", "number_type": "integer", "default": "20", "range": {"min": 0, "max": 40}}, {"name": "min_length", "label": "Minimal Sequence Length", "description": "Discard reads that became shorter than length INT because of either quality or adapter trimming. (Default: 20)", "type": "number", "number_type": "integer", "default": "20", "range": {"min": 0, "max": "Inf"}}, {"name": "additional_parameter", "label": "Additional Trim-Galore Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}, {"name": "gunzip", "installed": true}]}, "analyses": {"groupName": "analyses", "modules": [{"name": "deseq2", "installed": true, "settings": {"required": [{"name": "gff_feature_type", "label": "GFF Feature Type", "description": "Used feature type, e.g. gene or exon", "type": "string"}, {"name": "gff_feature_name", "label": "GFF Feature Name", "description": "Descriptor for gene name, e.g. ID or gene_id", "type": "string"}, {"name": "gff_path", "label": "GFF File Path", "description": "File path to gff file", "type": "file"}], "optional": [{"name": "additional_featcounts_options", "label": "Additional FeatureCounts Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}, {"name": "attribute_columns", "label": "GFF Attributes in Summary", "description": "GFF attributes to show in the beginning of the xlsx summary (Comma-separated list, e.g. \"experiment, product, Dbxref\")", "type": "string", "default": ""}], "columns": [{"name": "condition", "label": "Condition", "description": "Condition name of sequencing run", "type": "string"}]}}, {"name": "count_table", "installed": true, "settings": {"required": [{"name": "gff_feature_type", "label": "GFF Feature Type", "description": "Used feature type, e.g. gene or exon", "type": "string"}, {"name": "gff_feature_name", "label": "GFF Feature Name", "description": "Descriptor for gene name, e.g. ID or gene_id", "type": "string"}, {"name": "gff_path", "label": "GFF File Path", "description": "File path to gff file", "type": "file"}], "optional": [{"name": "additional_options", "label": "Additional FeatureCounts Options", "description": "Additional options to use in shell command", "type": "string", "default": ""}]}}, {"name": "readxplorer", "installed": true, "settings": {"required": [{"name": "readxplorer_cli_path", "label": "ReadXplorer CLI Path", "description": "Path to ReadXplorer CLI executable", "type": "file"}, {"name": "reference_genome", "label": "Reference Genome", "description": "Path to reference genome sequence", "type": "file"}]}}]}}